operators/pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -0,0 +1,301 @@
+	//  trigger restarts but just allow the sset controller to do it at its own pace.
+	//  Instead of green health, we could look at shards status, taking into account nodes
+	//  we scheduled for a restart (maybe not restarted yet).
+
+	maxConcurrentUpgrades := 1
+	scheduledUpgrades := 0
+
+	// Only update 1 master node at a time, for safety and zen settings convenience.
+	// This can slow down the upgrade, but the number of master nodes should be small anyway.
+	maxMasterNodeUpgrades := 1
+	scheduledMasterNodeUpgrades := 0
+
+	for i, statefulSet := range statefulSets {
+		// Inspect each pod, starting from the highest ordinal, and decrement the partition to allow
+		// pod upgrades to go through, controlled by the StatefulSet controller.
+		for partition := sset.GetUpdatePartition(statefulSet); partition >= 0; partition-- {
+			if partition >= sset.Replicas(statefulSet) {
+				continue
+			}
+			if scheduledUpgrades >= maxConcurrentUpgrades {
+				return results.WithResult(defaultRequeue)
+			}
+			if label.IsMasterNodeSet(statefulSet) && scheduledMasterNodeUpgrades >= maxMasterNodeUpgrades {
+				return results.WithResult(defaultRequeue)
+			}
+
+			// Do we need to upgrade that pod?
+			podName := sset.PodName(statefulSet.Name, partition)
+			podRef := types.NamespacedName{Namespace: statefulSet.Namespace, Name: podName}
+			alreadyUpgraded, err := podUpgradeDone(d.Client, esState, podRef, statefulSet.Status.UpdateRevision)
+			if err != nil {
+				return results.WithError(err)
+			}
+			if alreadyUpgraded {
+				continue
+			}
+
+			// An upgrade is required for that pod.
+			scheduledUpgrades++
+
+			// Is the pod upgrade already scheduled?
+			if partition == sset.GetUpdatePartition(statefulSet) {
+				continue
+			}
+
+			// Is the cluster ready for the node upgrade?
+			clusterReady, err := clusterReadyForNodeRestart(d.ES, esState)
+			if err != nil {
+				return results.WithError(err)
+			}
+			if !clusterReady {
+				// retry later
+				return results.WithResult(defaultRequeue)
+			}
+
+			log.Info("Preparing cluster for node restart", "namespace", d.ES.Namespace, "es_name", d.ES.Name)
+			if err := prepareClusterForNodeRestart(esClient, esState); err != nil {
+				return results.WithError(err)
+			}
+
+			if label.IsMasterNodeSet(statefulSet) {
+				scheduledMasterNodeUpgrades++
+				// TODO if the node is a master:
+				//  - zen1: update minimum_master_node to account for master node deletion. Otherwise upgrading a 2-masters
+				//   cluster provokes downtime since m_m_n=2.
+				//   Problem: how to prevent this to be reverted at the next reconciliation, before the pod gets deleted?
+				//  - zen2: set voting config exclusions: same problem, this is not easy. But since we only delete
+				//   one master at a time, maybe it's not required?
+			}
+
+			// Upgrade the pod.
+			if err := d.upgradeStatefulSetPartition(&statefulSets[i], partition); err != nil {
+				return results.WithError(err)
+			}
+			scheduledUpgrades++
```

Comment:
Should this be done twice per loop?

Reply:
Clearly not, good catch! (Also definitely shows this lacks some unit testing)